### ISSUE TITLE
Middleman: Message routing

### ISSUE THREAD

**Issue by cameel** [golemfactory/concent#629]
This is a part of the implementation of the "Concent Signing Service" blueprint (#597). This task depends on #618.

#### Overview
The purpose of this task is to make Middleman deliver messages from Concent to the Signing Service and route responses back over the right connection.

At this point we're not concerned with making operations atomic and responding with `ERROR` frames to invalid input. Messages are silently dropped in case of a crash or if they're invalid. We're going to deal with these problems properly in a separate issue.

##### Request Queue
This is [an asyncio Queue](https://docs.python.org/3/library/asyncio-queue.html).
- The queue is created when the Middleman starts, before any connection is established. It's never destroyed. This is because the messages in the queue must survive a connection drop. The only case where it's OK to destroy and recreate this queue is if the whole server managing connections to Concent crashes and is restarted.
- Request Producer can add items, Request Consumer can pop items.

Every item in the queue consists of:
- _connection ID_
- _Concent request ID_
- _message_
- _timestamp_

##### Response Queue pool
This is a mapping (`dict`) of [asyncio Queue](https://docs.python.org/3/library/asyncio-queue.html)s.
- Each queue identified by a connection ID.
- The pool is intially empty.
- A new queue is created for a Response Consumer and destroyed when the consumer stops or fails.
- Response Producer can add items to any queue. A single Response Consumer can pop items from its own queue.

Every item in the queue consists of:
- _message_
- _Concent request ID_
- _timestamp_

##### Message Tracker
A mapping between messages and connections. Should use the **Signing Service connection ID**s from the wire protocol. The mapping should preserve order in which items are added so that the Response Producer can process them in that order.

Request Consumer can add messages. Response Producer can remove them.

Each entry consists of:
- _Signing Service request ID_
                            ||
                            V
- _Concent request ID_
- _connection ID_
- _message_
- _timestamp_

##### Concent connection handler
This replaces the handler described in #615:

1. Create a single Response Queue.
2. Create a Request Producer, a Response Consumer.
    - Producer should get the socket reader, and consumer should get the writer. They take ownership over them and are responsible for closing and destroying them.
    - The handler owns the producer and consumer. Both should be stopped and destroyed if the handler ends of fails.
3. If either the consumer or the producer fails, the whole handler fails. It should clean up after its coroutines and reraise the exception so that the connection gets interrupted. The server, not the handler is responsible for reporting and silencing the exception.
    - When a queue is destroyed, log information about each dropped message at `INFO` level.
4. When the client closes the connection, exit without errors.

##### Signing Service connection handler
This replaces the handler described in #616.

This handler is very similar to the Concent connection handler. The only differences are:
- If a connection is already established close the connection and return. There can be only one connection with the service.
- It should not create or destroy any queue.
- The producer and consumer it creates are the Response Producer and the Request Consumer.

##### Request producer
Do this in a loop until the connection drops or the coroutine is stopped by its owner:
1. Read a single, complete message from Concent.
    - If the message is incomplete or invalid, drop it. Log the issue at `INFO` level.
2. Log at `INFO` level that a request has been received. Include the request ID and connection ID.
3. Add the message to the Request Queue along with connection ID and its request ID (Concent request ID).

##### Request consumer
Do this in a loop until the connection drops or the coroutine is stopped by its owner:
1. Pop a single item from the Request Queue.
2. If the Response Queue corresponding to the connection ID no longer exists, drop the message. Log the issue at `INFO` level. Then end the current pass of the loop.
3. Generate a new, unique request ID (Signing Service request ID).
    - You can create an entirely new one or just use Concent request ID which is guaranteed to be unique within a single connection and combine it with connection ID.
4. Add the item to the Message Tracker along with all the IDs.
5. Send the message to the Signing Service.

##### Response producer
Do this in a loop until the connection drops or the coroutine is stopped by its owner:
1. Read a single, complete message from the Signing Service.
2. Use request ID from the received message to look up an entry with matching Signing Service request ID in the Message Tracker.
3. If there's no such entry, drop the message. Log the issue at `INFO` level. Then end the current pass of the loop.
4. If the Response Queue corresponding to the connection ID no longer exists, drop the message. Log the issue at `INFO` level. Then end the current pass of the loop.
5. Discard all entries from the Message Tracker that were added before the one corresponding to the message we have just received. Log information about each dropped message at `INFO` level. 
6. Add the message to the right Response Queue along with the Concent request ID.

##### Response consumer
Do this in a loop until the connection drops or the coroutine is stopped by its owner:
1. Pop a single item from the Request Queue.
2. Send the message to Concent
3. Log at `INFO` level that the message as been delivered successfully.

**Comment by PaweuB** [comment]
Pivotal:
https://www.pivotaltracker.com/story/show/158999817

**Comment by dybi** [comment]
I have edited the issue description; mostly data structures are simplified.
Below the old issue description for reference:
--------------------------------
This is a part of the implementation of the "Concent Signing Service" blueprint (#597). This task depends on #618.

#### Overview
The purpose of this task is to make Middleman deliver messages from Concent to the Signing Service and route responses back over the right connection.

At this point we're not concerned with making operations atomic and responding with `ERROR` frames to invalid input. Messages are silently dropped in case of a crash or if they're invalid. We're going to deal with these problems properly in a separate issue.

#### Global state
##### Request Queue
This is [an asyncio Queue](https://docs.python.org/3/library/asyncio-queue.html).
- The queue is created when the Middleman starts, before any connection is established. It's never destroyed. This is because the messages in the queue must survive a connection drop. The only case where it's OK to destroy and recreate this queue is if the whole server managing connections to Concent crashes and is restarted.
- Request Producer can add items, Request Consumer can pop items.

Every item in the queue consists of:
- _connection ID_
- _Concent request ID_
- _message_
- _timestamp_

##### Response Queue pool
This is a set of [asyncio Queue](https://docs.python.org/3/library/asyncio-queue.html)s.
- Each queue identified by a connection ID.
- The pool is intially empty.
- A new queue is created for a Response Consumer and destroyed when the consumer stops or fails.
- Response Producer can add items to any queue. A single Response Consumer can pop items from its own queue.

Every item in the queue consists of:
- _message_
- _Concent request ID_
- _timestamp_

##### Message Tracker
A mapping between messages and connections. Should use the **frame ID**s from the wire protocol. The mapping should preserve order in which items are added so that the Response Producer can process them in that order.

Request Consumer can add messages. Response Producer can remove them.

Each entry consists of:
- _Signing Service request ID_
- _Concent request ID_
- _connection ID_
- _message_
- _timestamp_

#### Handlers
##### Concent connection handler
This replaces the handler described in #615:

1. Create a single Response Queue.
2. Create a Request Producer, a Response Consumer.
    - Producer should get the socket reader, and consumer should get the writer. They take ownership over them and are responsible for closing and destroying them.
    - The handler owns the producer and consumer. Both should be stopped and destroyed if the handler ends of fails.
3. If either the consumer or the producer fails, the whole handler fails. It should clean up after its coroutines and reraise the exception so that the connection gets interrupted. The server, not the handler is responsible for reporting and silencing the exception.
    - When a queue is destroyed, log information about each dropped message at `INFO` level.
4. When the client closes the connection, exit without errors.

##### Signing Service connection handler
This replaces the handler described in #616.

This handler is very similar to the Concent connection handler. The only differences are:
- If a connection is already established close the connection and return. There can be only one connection with the service.
- It should not create or destroy any queue.
- The producer and consumer it creates are the Response Producer and the Request Consumer.

##### Request producer
Do this in a loop until the connection drops or the coroutine is stopped by its owner:
1. Read a single, complete message from Concent.
    - If the message is incomplete or invalid, drop it. Log the issue at `INFO` level.
2. Log at `INFO` level that a request has been received. Include the request ID and connection ID.
3. Add the message to the Request Queue along with connection ID and its request ID (Concent request ID).

##### Request consumer
Do this in a loop until the connection drops or the coroutine is stopped by its owner:
1. Pop a single item from the Request Queue.
2. If the Response Queue corresponding to the connection ID no longer exists, drop the message. Log the issue at `INFO` level. Then end the current pass of the loop.
3. Generate a new, unique request ID (Signing Service request ID).
    - You can create an entirely new one or just use Concent request ID which is guaranteed to be unique within a single connection and combine it with connection ID.
4. Add the item to the Message Tracker along with all the IDs.
5. Send the message to the Signing Service.

##### Response producer
Do this in a loop until the connection drops or the coroutine is stopped by its owner:
1. Read a single, complete message from the Signing Service.
2. Use request ID from the received message to look up an entry with matching Signing Service request ID in the Message Tracker.
3. If there's no such entry, drop the message. Log the issue at `INFO` level. Then end the current pass of the loop.
4. If the Response Queue corresponding to the connection ID no longer exists, drop the message. Log the issue at `INFO` level. Then end the current pass of the loop.
5. Discard all entries from the Message Tracker that were added before the one corresponding to the message we have just received. Log information about each dropped message at `INFO` level. 
6. Add the message to the right Response Queue along with the Concent request ID.

##### Response consumer
Do this in a loop until the connection drops or the coroutine is stopped by its owner:
1. Pop a single item from the Request Queue.
2. Send the message to Concent
3. Log at `INFO` level that the message as been delivered successfully.
